src/client.js

```diff
@@ -0,0 +1,351 @@
+ * {
+ *   username:          // Pulse username
+ *   password:          // Pulse password
+ *   hostname:          // Hostname to use
+ * }
+ */
+const buildConnectionString = function({username, password, hostname}) {
+  assert(username, 'options.username is required');
+  assert(password, 'options.password is required');
+  assert(hostname, 'options.hostname is required');
+
+  // Construct connection string
+  return [
+    'amqps://',         // Ensure that we're using SSL
+    encodeURI(username),
+    ':',
+    encodeURI(password),
+    '@',
+    hostname,
+    ':',
+    5671,                // Port for SSL
+  ].join('');
+};
+exports.buildConnectionString = buildConnectionString;
+
+/**
+ * An object to create connections to a pulse server.  This class will
+ * automatically handle reconnecting as necessary.
+ *
+ * AMQP is a very connection-oriented protocol.  For example, a client using
+ * non- durable queues will need to re-declare those queues on every new
+ * connection.  Similarly, a consumer must re-start consumption on every new
+ * connection.  This class emits a `connected` event on each new
+ * connection, and that function should re-establish any state as required for
+ * the new connection.
+ *
+ * Connections are automatically cycled periodically, regardless of any problems
+ * with the connection itself, in order to exercise the reconnection logic. When
+ * this occurs, the old connection is held open for 30 seconds to allow any pending
+ * publish operations or message consumptions to complete.
+ *
+ * Options:
+ * * connectionString
+ * * username
+ * * password
+ * * hostname
+ * * recycleInterval (ms; default 1h)
+ * * retirementDelay (ms; default 30s)
+ * * minReconnectionInterval (ms; default 15s)
+ * * monitor (taskcluster-lib-monitor instance)
+ *
+ * The pulse namespace for this user is available as `client.namespace`.
+ */
+class Client extends events.EventEmitter {
+  constructor({username, password, hostname, connectionString, recycleInterval,
+    retirementDelay, minReconnectionInterval, monitor}) {
+    super();
+
+    if (connectionString) {
+      assert(!username, 'Can\'t use `username` along with `connectionString`');
+      assert(!password, 'Can\'t use `password` along with `connectionString`');
+      assert(!hostname, 'Can\'t use `hostname` along with `connectionString`');
+      this.connectionString = connectionString;
+
+      // extract the username as namespace
+      const connURL = new URL(connectionString);
+      this.namespace = decodeURI(connURL.username);
+    } else {
+      this.connectionString = buildConnectionString({username, password, hostname});
+      this.namespace = username;
+    }
+
+    assert(monitor, 'monitor is required');
+    this.monitor = monitor;
+
+    this.recycleInterval = recycleInterval || 3600 * 1000;
+    this.retirementDelay = retirementDelay || 30 * 1000;
+    this.minReconnectionInterval = minReconnectionInterval || 15 * 1000;
```

Comment:
I think that these should be made private with a `_` so people don't unwilling read these properties..

this goes for a lot of the properties and methods in all classes... If things are not intented to be public prefixing `_` is a good way to go.